### PR TITLE
Support for apply from directly in the buildscript block

### LINKED ISSUE #497: "apply from" support in buildscript block
I am trying to convert `build.gradle` into kotlin-dsl:
```
buildscript { 
  apply from: "gradle/buildscript.gradle", to: buildscript 
}
apply from: "gradle/project-footer.gradle"
```

The following `build.gradle.kts` gives me `build.gradle.kts:3:23: Unresolved reference: from`:
```
import org.gradle.kotlin.dsl.apply

buildscript { 
  apply { 
    from("gradle/buildscript.gradle")
    to(buildscript) 
  } 
}
apply { 
  from("gradle/project-footer.gradle")
}
```

It seems that the "apply from" support in the buildscript block is missing.

Ref. https://kotlinlang.slack.com/archives/C19FD9681/p1505817768000291

### PR DESCRIPTION
**Pull request by JLLeitschuh**
Before you had do to this:
```kotlin
buildscript {
    project.apply {
        from("repositoryConfig.gradle.kts")
    }
}
```
Now you can do this:
```kotlin
buildscript {
    apply {
        from("repositoryConfig.gradle.kts")
    }
}
```

This improves the parity with the Groovy DSL

Closes #497

### Contributor Checklist
- [x] Base the PR against the `develop` branch
- [x] [Sign Gradle CLA](http://gradle.org/contributor-license-agreement/)
- [x] Provide integration tests to verify changes from a user perspective
- [x] Provide unit tests to verify logic
- [x] Ensure that tests pass locally: `./gradlew check --parallel`

### CODE REVIEW

**Comment by JLLeitschuh** [comment]
@bamboo I can't request you as a reviewer. Can you have a look at this?

**Comment by JLLeitschuh** [comment]
@bamboo @eskatos Do either of you have any thoughts on this? Do you want this?

**Comment by bamboo** [comment]
Thanks, @JLLeitschuh, we definitely want to have it. I'll get back to it soon.

**Comment by bamboo** [comment]
This problem arises because Kotlin prefers methods applicable to the innermost receiver over methods available to outer receivers (such as the build script itself and its enclosing project) and the Kotlin standard library already defines an extension method named `apply` available to any receiver:

```kotlin
inline fun <T> T.apply(block: T.() -> Unit): T
```

Which ends up shadowing [`PluginAware.apply(Action<in ObjectConfigurationAction>)`](https://docs.gradle.org/current/javadoc/org/gradle/api/plugins/PluginAware.html#apply-org.gradle.api.Action-).

This also means the problem is not confined to the `buildscript` block. Any block that implies a receiver that is not `PluginAware` is affected, for instance:

```kotlin
dependencies {
  apply { from("my-dependencies.gradle.kts") } // Unresolved reference: from
}
```

In other words, any invocation of the form `apply { <lambda> }` would match the standard library `apply` extension unless the innermost receiver is `PluginAware`.

Because of that, an adhoc solution that only mitigates this problem in the context of the `buildscript` block doesn't seem satisfactory.

We need a holistic strategy before we can move on with this PR.

**Comment by henrik242** [comment]
@bamboo This also applies to `apply { to() }`, right? (Ref. https://github.com/gradle/kotlin-dsl/issues/497 )  I'm currently looking for a workaround while waiting for a proper fix, any ideas?

**Comment by bamboo** [comment]
@henrik242, yes.

**Comment by bamboo** [comment]
Superseded by #659 and #682.
